### DISCUSSION TITLE
Arduino R4 Wifi, renesas_uno/1.0.5 - can't find header files?

### DISCUSSION THREAD

**Discussion by staffanbruce** [Sloeber/arduino-eclipse-plugin#1600] [Q&A] [answered]
Hello,

I'm trying to compile a simple example file for the Arduino Uno R4 WIFI, but run into problems immediately. 
Include file bsp_api.h and several others are not found although they exist within the board definition path.

```
21:13:44 **** Build of configuration Release for project R4_WIFI ****
"D:\\Program\\Sloeber\\arduinoPlugin\\tools\\make\\make" all 
'Building file: ..\sloeber.ino.cpp'
'Starting C++ compile'
"D:\Program\Sloeber\arduinoPlugin\packages\arduino\tools\arm-none-eabi-gcc\7-2017q4/bin/arm-none-eabi-g++" -c -Wall -Wextra -Os -g3 -fno-use-cxa-atexit -fno-rtti -fno-exceptions -MMD -nostdlib -DF_CPU=48000000 -DNO_USB -DBACKTRACE_SUPPORT -DARDUINO_UNOR4_WIFI -MMD -std=gnu++17 -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsigned-char -ffunction-sections -fdata-sections -fmessage-length=0 -fno-builtin -DARDUINO=10802 "-DPROJECT_NAME="D:/sloeber-workspace/R4_WIFI/Release/R4_WIFI"" -DARDUINO_UNOWIFIR4 -DARDUINO_ARCH_RENESAS_UNO -DARDUINO_ARCH_RENESAS -DARDUINO_FSP -D_XOPEN_SOURCE=700 -mthumb "@D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\variants\UNOWIFIR4/defines.txt" -DCFG_TUSB_MCU=OPT_MCU_RAXXX  "-ID:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\cores\arduino/tinyusb" "-ID:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\cores\arduino/api/deprecated" "-ID:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\cores\arduino/api/deprecated-avr-comp"  "-iprefixD:\Program\Eclipse\2019-09/arduinoPlugin/packages/arduino/hardware/renesas_uno/1.0.5" "@D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\variants\UNOWIFIR4/includes.txt"  -I"D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\cores\arduino" -I"D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\variants\UNOWIFIR4" -I"D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\libraries\Arduino_LED_Matrix\src" -MMD -MP -MF"sloeber.ino.cpp.d" -MT"sloeber.ino.cpp.o" -D__IN_ECLIPSE__=1 -x c++ "..\sloeber.ino.cpp"  -o  "sloeber.ino.cpp.o"
In file included from ..\sloeber.ino.cpp:7:0:
D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\cores\arduino/Arduino.h:5:10: fatal error: bsp_api.h: No such file or directory
 #include "bsp_api.h"
          ^~~~~~~~~~~
compilation terminated.
```

The above compilation is based on code starting from scratch 
* Arduino/Create new sketch
* Select Platform folder .../arduino/hardware/renesas_uno/1.05
* Select board Arduino UNO R4 Wifi
* Select a code example Library/Arduino_LED_Matrix/GameOfLife. The example code can be seemingly any alternative. All give the same error.

The file "bsp_api.h" exists under: ...\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\variants\UNOWIFIR4\includes\ra\fsp\inc\api

Should/can I in some way add this path or is there another way? There is no specific folder called bsp_api in the libraries folder so there is no way of selecting this header/library under Arduino/Add a library to the selected project. 
Is there a way to initiate a search for headers in some way? How should Sloeber be made aware of these libraries?

There seem to be a post "How Sloeber deals with libraries; the library manager; import into project and install #1537" discussing similar things, but I don't see what's missing in my configuration.

Any help would be very much appreciated. 

Regards,
 Staffan

**Comment by jantje** [comment]
I seem to be unable to find the board you use
.\arduinoPlugin\packages\arduino\hardware\ **renesas_uno** \1.0.5\variants\
![afbeelding](https://github.com/Sloeber/arduino-eclipse-plugin/assets/1175841/0fb403ba-dd2e-48c2-a443-af690cbd54dc)

Can you tell me which json you used to download this package?
Can you provide a image of the project properties->sloeber page?

**Reply by staffanbruce** [reply]
Hi,

Not sure I have the exact answers, but here goes.

The boards are defined/selected through Eclipse Arduino->Preferences->Platforms and boards:
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/ddb136ec-d0c0-40bf-b8e1-66f5ec6b2c75)

Not sure I have the project properties->sloeber page, but Eclipse Arduino->Preferences->Arduino:
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/23a33153-5606-4230-b1eb-d83185acaa9f)

Was this what you were looking for?

The folder where the board definitions are stored looks like this:
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/d3d646d8-b136-4d48-8c8a-b70f16a9db93)


Regards,
  Staffan

**Reply by jantje** [reply]
>The boards are defined/selected through Eclipse Arduino->Preferences->Platforms and boards:

The boards you show there have a version 1.0.2 and 1.0.1 where the board you use is a 1.0.5. So this is not the way you installed the boards.

>Not sure I have the project properties->sloeber page, but Eclipse Arduino->Preferences->Arduino:

This is not what I asked for. But:
Remove the C:/users/... stuff from the private hardware and library path. Who told you to do this?
What I asked is: From the project explorer right click on the project and select properties->sloeber

>The folder where the board definitions are stored looks like this:

I know: What I do not know is: How was this folder created. I'm not aware of a platform installable via a json that creates a path like that.
You obviously followed some documentation. Can you provide me these instructions because honestly .... I would strongly advice not to follow these instructions.

**Reply by staffanbruce** [reply]
Hi,

> The boards you show there have a version 1.0.2 and 1.0.1 where the board you use is a 1.0.5. So this is not the way you installed the boards.

The boards are there, only some 'help' window hiding 1.0.4 and 1.0.5
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/5d0b1ba9-3141-4f34-989e-6c48876cfd2e)

> From the project explorer right click on the project and select properties->sloeber

I tried to find the sloeber section, but I get this when right clicking on the project and select properties:
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/413f3ee3-f7e9-44cf-a050-35bdc4ed987b)

> I know: What I do not know is: How was this folder created. I'm not aware of a platform installable via a json that creates a path like that.

Sorry, no real clue honestly. I selected the board as described and some magic happened... I really wish I knew more about what is going on under the hood - but I don't. 
I believe I found the path by figuring it should be somewhere in the install path of Sloeber...

> You obviously followed some documentation. Can you provide me these instructions because honestly .... I would strongly advice not to follow these instructions.

Trial and error mainly - no documentation. 

Looking at it from another angle. Is there any way to install a definition for the Arduino R4 Wifi? I just thought this would be the way to do it, but if anyone can describe a better way, that would really be what I'm looking for - being able to write programs for these boards.

Regards,
  Staffan

**Reply by jantje** [reply]
>The boards are there, only some 'help' window hiding 1.0.4 and 1.0.5

It seems something went wrong here on my side and that indeed is the info I was looking for.

>I tried to find the sloeber section, but I get this when right clicking on the project and select properties:

That is it indeed. It is called arduino because you are not using the latest sloeber.

>Sorry, no real clue honestly. I selected the board as described and some magic happened... 

That is how it is supposed to be.

Anyway: Now I have the board I tried here locally and I had some issues. I tried in Arduino IDE and that worked fine.
So I had another look at how Sloeber did it.
There were 2 issues
![afbeelding](https://github.com/Sloeber/arduino-eclipse-plugin/assets/1175841/58be3e5c-4933-4128-bd63-c3a9c8a72588)

1)For one reason or another no `#pragma once` was added to the Arduino_LED_Matrix.h file. You can add that yourself like I did.
2) A whole bunch of libraries were added to the project. Just delete the libraries folder. Try to build a couple of times (deleting the libraries folder if unneeded libraries are added). Mostly Sloeber gets it right after 3 build trials. It is a manually recursive process ;-).

Then it worked fine.
As you are new to sloeber 
1) can you use this tutorial and provide feedback? https://github.com/Sloeber/arduino-eclipse-plugin/discussions/1342
2) Here is a document describing how Sloeber selects libraries.
3) If libraries keep popping up you can manually add libraries or use (right click the library folder) exclude from build. I did a write up on this somewhere but I fail  to find it.

**Reply by staffanbruce** [reply]
Hello,

Not sure I get it working.
I tried adding the "#pragma once":
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/17b7b6d2-c1f1-40eb-ad59-4ff3be2ab34e)

...but then, I'm reluctant to delete the libraries folder since the Arduino_LED_Matrix.h resides there. Also, it seems like deleting the folder actually removes the files fundamentally from the installed folder, not just my project folder? Here is what properties says about Arduino_LED_Matrix library;
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/bfc3d9e8-fe53-4bc7-9d26-6361c612713f)

So, with "#pragma once" added, but not deleting any libraries (yet), I still get the same error.

There seem to be no other libraries added in addition to the Arduino_LED_Matrix - am I missing something?

> can you use this tutorial and provide feedback? https://github.com/Sloeber/arduino-eclipse-plugin/discussions/1342

Great thing! I have browsed through and found some very good stuff. Suggest changing the "Change platform to ESP32" into a generic "Change platform" and just clarify that Esp32 is used as an example. I figure the json-files you referred to earlier in this thread are ones like the one described in the tutorial (//raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_index.json)? This was great info.
Two things that I couldn't find (easily?): 
* How to add separate cpp-files. Knowing that it shall be .cpp and .h rather than (old fashioned) .c and .h took me a while. 
* How to migrate to newer versions of Sloeber. Since I have a few projects running, I'm very hesitant to change anything that works (if not broken - don't fix...). Understanding how to change version without breaking the IDE and Git etc - knowing what will happen and what may be deleted/lost would be so good. 

Perhaps this information is readily available - perhaps just a link in the document?

I have downloaded the word document. At least the ambition is to read through and comment further. 

> Here is a document describing how Sloeber selects libraries.

That would be very good to read also - link missing?

> If libraries keep popping up you can manually add libraries or use (right click the library folder) exclude from build. I did a write up on this somewhere but I fail to find it.

Is this the Resource Configurations->Exclude from build? What would cause the addition of unwanted libraries? Are there specific ones to look out for?
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/49ad82a0-501a-4a1d-a837-1c6ebdd2fc4b)

Not completely home, but seems to be getting close. It would be so good to be able to compile!

Regards,
  Staffan

**Reply by jantje** [reply]
> ...but then, I'm reluctant to delete the libraries folder since the Arduino_LED_Matrix.h resides there. 

No worries. It is all linked stuff. You can see it is a linked folder by the small arrow 
![afbeelding](https://github.com/Sloeber/arduino-eclipse-plugin/assets/1175841/c0fcd5da-ac24-4941-8eec-f7536dc529f8)
So nothing gets deleted from the file system.
But as you only have the 1 library there is no need to delete the folder.


>So, with "#pragma once" added, but not deleting any libraries (yet), I still get the same error.

I took a deeper look here how Sloeber finds this bsp_api.h file.
If you look at the command line you provided you wiill find this part.
"@D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\variants\UNOWIFIR4/includes.txt"
This basically states that the file includes.txt contains more command line options.
D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\variants\UNOWIFIR4 should be linked to your core/variant folder.
so the content core/variant/includes.txt  is added to the command line.
On my system includes.txt contains
`-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/inc/api`
The gnu doc states https://gcc.gnu.org/onlinedocs/gcc-4.4.2/gcc/Preprocessor-Options.html
>-iwithprefixbefore dir
>    Append dir to the prefix specified previously with -iprefix, and add the resulting directory to the include search path. -iwithprefixbefore puts it in the same place -I would; -iwithprefix puts it where -idirafter would. 

as your command contains
`"-iprefixD:\Program\Eclipse\2019-09/arduinoPlugin/packages/arduino/hardware/renesas_uno/1.0.5"`
That seems fine to me.
I do not understand why The compiler does not find the bsp_api.h file.
If you ctrl click the `#include "bsp_api.h"` does the file get opened?

>Is this the Resource Configurations->Exclude from build? What would cause the addition of unwanted libraries? Are there specific ones to look out for?

The most common case is with boards with their own wifi/ethernet taking in common ethernet libraries.

**Reply by staffanbruce** [reply]
> If you ctrl click the #include "bsp_api.h" does the file get opened?

Yes it does - and it is also under core:

![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/0d74167c-2eff-425a-8b43-4810f2ec83f8)

Although, hovering over the file in Arduino.h I get this:
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/2429d8ea-12cc-4d67-a368-67f1cf15b10d)


The file includes.txt looks like this:
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/inc
**-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/inc/api**
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/inc/instances
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/arm/CMSIS_5/CMSIS/Core/Include
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra_gen
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra_cfg/fsp_cfg/bsp
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra_cfg/fsp_cfg
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/src/r_usb_basic/src/driver/inc
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/src/r_sce/crypto_procedures/src/sce5/plainkey/private/inc
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/src/r_sce/crypto_procedures/src/sce5/plainkey/public/inc
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/src/r_sce/crypto_procedures/src/sce5/plainkey/primitive
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/src/r_sce/common
-iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/src/r_sce


Noting that -iwithprefixbefore/variants/UNOWIFIR4/includes/ra/fsp/inc/api indeed seems to point to the folder... ???

I guess what you point out is that with the following being part of the command line:

"-iprefixD:\Program\Eclipse\2019-09/arduinoPlugin/packages/arduino/hardware/renesas_uno/1.0.5" "@D:\Program\Sloeber\arduinoPlugin\packages\arduino\hardware\renesas_uno\1.0.5\variants\UNOWIFIR4/includes.txt" 

...it should look in 

D:\Program\Eclipse\2019-09/arduinoPlugin/packages/arduino/hardware/renesas_uno/1.0.5/variants/UNOWIFIR4/includes/ra/fsp/inc/api

...but there is no folder arduinoPlugin under the folder D:\Program\Eclipse\2019-09. The folder arduinoPlugin resides in D:\Program\Sloeber.
Is there a confusion in configurations somewhere? 

Regards,
  Staffan

**Reply by jantje** [reply]
Seems you need the very latest sloeber #1448
The platform.txt states

```
## Compile c files
recipe.c.o.pattern="{compiler.path}{compiler.c.cmd}" {compiler.c.flags} -DARDUINO={runtime.ide.version} -DARDUINO_{build.board} -DARDUINO_ARCH_{build.arch} -DARDUINO_ARCH_RENESAS -DARDUINO_FSP -D_XOPEN_SOURCE=700 {compiler.fsp.cflags} {compiler.tinyusb.cflags}  {build.extra_flags} {tinyusb.includes} "-I{build.core.path}/api/deprecated" "-I{build.core.path}/api/deprecated-avr-comp" {includes} "-iprefix{runtime.platform.path}" "@{compiler.fsp.includes}" -o "{object_file}" "{source_file}"
```

Note the 
"-iprefix{runtime.platform.path}"
And runtime.platform.path was wrong before #1448 

>How to migrate to newer versions of Sloeber. 
Unless you have loots of added plugins I advice the install product next to current project.
http://eclipse.baeyens.it/how_to.shtml#/n

**Reply by staffanbruce** [reply]
Thanks for the link on how to upgrade.
I have downloaded and run 4.4.2, but run into problem when trying the restore configuration of the project. Selecting properties->Sloeber, I get the following:
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/93089932-c4fd-44b0-b461-9d04d937f10c)


When starting a new project I can select the right board (it seems) and the Sloeber window appears OK:
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/2b3d7e75-66fe-4a27-a66f-3b707332a7c4)

Also, now I don't see the output on the console when compiling anymore. It is totally blank so not clear what is going on there.

What I have done:
1. Downloaded Sloeber 4.4.2 https://eclipse.baeyens.it/stable.php?OS=Windows
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/057b22d2-a993-4257-852e-cff74df3d6e8)
2. Copied content to d:\Program\Sloeber_4.4.2
3. Backed up D:\sloeber-workspace
4. Started Eclipse/Sloeber via this link:
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/f5908738-92a2-49ac-ad3d-6b7ba862b8c8)


Any ideas?

Regards,
  Staffan

**Reply by staffanbruce** [reply]
Should add: no splash screen when starting anymore. Perhaps standard.

**Reply by jantje** [reply]
First thing to notice:
Sloeber stores its package data in the arduinPlugin folder under the eclipse install (or SLOEBER_HOME variable location).
Using a new sloeber means you have to reinstall (or copy the arduinoplugin folder) all the boards.
Second thing to not(ice
You can always try "file->new->other->convert to  arduino project" to convert old sloeber projects
![afbeelding](https://github.com/Sloeber/arduino-eclipse-plugin/assets/1175841/9c1b7d40-0df3-4818-85fd-00796ade9507)

>now I don't see the output on the console when compiling anymore.
Are you in sloeber or C/C++ perspecive?
Have you closed the console and restarted a build?

>no splash screen when starting anymore. 
Not sure what is going on on your system but there is a splash screen issue.

**Reply by staffanbruce** [reply]
Partial success?
Instead of running 4.4.2, I run 4.4.1. 

>You can always try "file->new->other->convert to arduino project" to convert old sloeber projects

This works in 4.4.1 and I can update the information in project->Properties->Sloeber. A little unclear which "Programmer protocol" to use though. Haven't tested so far.

I also tried to continue to 4.4.2 after updating at 4.4.1, but still same issue with 
![image](https://github.com/Sloeber/arduino-eclipse-plugin/assets/6115043/3d1f2212-9d89-4f3f-b0d8-4d9f51778d12)

Back to 4.4.1, adding "#pragma once" to the file Arduino_LED_Matrix.h seems to do the trick here. The files I have tested so far compiles!

It's too bad I can't run 4.4.2. I'm running a plain vanilla Win10 computer. Are there any known issues?
Still, great to get it running under 4.4.1!

Regards,
  Staffan

**Reply by jantje** [reply, marked as answer]
>Instead of running 4.4.2, I run 4.4.1.

As the fix seems to be in 4.4.1 (I thought 4.4.2 but it seems to be 4.4.1) 4.4.1 is fine.
I recently released 4.4.2 because Sloeber didn't build and did not install into the latest eclipse via market place.
It may be there is regression in 4.4.2.

>"Programmer protocol" to use though.

If you do not own a programmer (if you own one you will know) just leave it blank.
4.4 introduced burning bootloader and a different button for upload and upload using programmer (like arduino IDE does)

**Reply by staffanbruce** [reply]
> As the fix seems to be in 4.4.1 (I thought 4.4.2 but it seems to be 4.4.1) 4.4.1 is fine.

Good to know - will stay there for now. Many thanks!

Regards,
  Staffan
